flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStorage.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+
+/**
+ * CheckpointStorage implements the durable storage of checkpoint data and metadata streams.
+ * An individual checkpoint or savepoint is stored to a {@link CheckpointStorageLocation},
+ * created by this class.
+ */
+public interface CheckpointStorage {
+
+	/**
+	 * Checks whether this backend supports highly available storage of data.
+	 *
+	 * <p>Some state backends may offer support for that with default settings, which makes them
+	 * suitable for zero-config prototyping, but not for actual production setups.
+	 */
+	boolean supportsHighlyAvailableStorage();
```

Comment:
The purpose of this did not become fully clear for me from the comments and the method is also never used. If this is part of some future work, you could consider adding the method in a future PR.

Reply:
Yes, this is part of future work. The high-level description says that:

> To make sure there is no extra persisting of the checkpoint metadata by the HA store (it simply references the regular persisted checkpoint metadata) we need some changes to the ZooKeeperCompletedCheckpointStore.

The method helps as a sanity check, to allow the checkpoint coordinator to validate that if the `CompletedCheckpointStore` requires durable persistence, then the `CheckpointStore` must provide that. It is used to catch a weird corner case where the ZkCompletedCheckpointStore is used, but the StateStore (from the MemoryStateBackend) does not support actually persisting any data durably. We could remove that, because currently this can never happen (The MemoryStateBackend automatically fills in durable persistence in HA setups)